CMakeLists.txt

```diff
@@ -99,83 +99,134 @@ if(FCL_COVERALLS)
 endif()
 
 find_package(PkgConfig QUIET)
+set(PKG_CONFIG_USE_CMAKE_PREFIX_PATH ON)
 
-# Find Eigen3
-find_package(Eigen3 3.0.5 QUIET)
-if(EIGEN3_FOUND)
+#===============================================================================
+# Find required dependency Eigen3 (>= 3.0.5)
+#
+# If Eigen3 is not found, manually set the cache variable EIGEN3_INCLUDE_DIR
+#===============================================================================
+find_package(Eigen3 3.0.5 QUIET CONFIG)
+
+# If Eigen3Config.cmake is not found, use the FindEigen3.cmake module
+if(NOT Eigen3_FOUND)
+  find_package(Eigen3 3.0.5 QUIET MODULE)
+  set(Eigen3_FOUND ON)
+endif()
+
+if(Eigen3_FOUND)
   set(FCL_HAVE_EIGEN TRUE)
-  include_directories(SYSTEM ${EIGEN3_INCLUDE_DIR})
 else()
   message(SEND_ERROR "EIGEN3 (>= 3.0.5) is required by FCL")
   set(FCL_HAVE_EIGEN FALSE)
 endif()
 
-# Find libccd
+#===============================================================================
+# Find required dependency libccd
+#
+# If libccd is not found, manually set the cache variables CCD_INCLUDE_DIR and
+# CCD_LIBRARY
+#===============================================================================
 find_package(ccd QUIET)
 
-if(NOT CCD_FOUND AND PKG_CONFIG_FOUND)
-  pkg_check_modules(CCD ccd)
-  # check to see if the pkg is installed under the libccd name
-  if(NOT CCD_FOUND)
-    pkg_check_modules(CCD libccd)
+# If ccd-config.cmake is not found, use pkg-config and/or find_path() and
+# find_library()
+if(NOT ccd_FOUND)
+  if(PKG_CONFIG_FOUND)
+    pkg_check_modules(PC_CCD ccd)
+    pkg_check_modules(PC_LIBCCD libccd)
+  endif()
+
+  find_path(CCD_INCLUDE_DIR ccd/ccd.h
+    HINTS "${PC_CCD_INCLUDE_DIRS}" "${PC_LIBCCD_INCLUDE_DIRS}")
+
+  # Using find_library() even if pkg-config is available ensures that the full
+  # path to the ccd library is available in CCD_LIBRARIES
+  find_library(CCD_LIBRARY ccd
+    HINTS "${PC_CCD_LIBRARY_DIRS}" "${PC_LIBCCD_LIBRARY_DIRS}")
+
+  # libccd links to LibM on UNIX.
+  if(CYGWIN OR NOT WIN32)
+    find_library(M_LIBRARY m)
+  endif()
+
+  if(CCD_INCLUDE_DIR AND CCD_LIBRARY)
+    set(CCD_INCLUDE_DIRS "${CCD_INCLUDE_DIR}")
+    set(CCD_LIBRARIES "${CCD_LIBRARY}" "${M_LIBRARY}")
+    set(ccd_FOUND ON)
+
+    mark_as_advanced(CCD_INCLUDE_DIR CCD_LIBRARY)
   endif()
 endif()
```

Comment:
I would like to suggest encapsulating this logic (and for other dependencies as well) as a separate custom module like `FindCCD.cmake`. That would make this main CMakeLists.txt cleaner.